zebra-test/src/transcript.rs

```diff
@@ -0,0 +1,81 @@
+//! A [`Service`](tower::Service) implementation based on a fixed transcript.
+
+use color_eyre::eyre::{ensure, eyre, Report};
+use futures::future::{ready, Ready};
+use std::{
+    fmt::Debug,
+    task::{Context, Poll},
+};
+use tower::{Service, ServiceExt};
+
+pub struct Transcript<R, S, I>
+where
+    I: Iterator<Item = (R, S)>,
+{
+    messages: I,
+}
+
+impl<R, S, I> From<I> for Transcript<R, S, I>
+where
+    I: Iterator<Item = (R, S)>,
+{
+    fn from(messages: I) -> Self {
+        Self { messages }
+    }
+}
+
+impl<R, S, I> Transcript<R, S, I>
+where
+    I: Iterator<Item = (R, S)>,
+    R: Debug,
+    S: Debug + Eq,
+{
+    pub async fn check<C>(mut self, mut to_check: C) -> Result<(), Report>
+    where
+        C: Service<R, Response = S>,
+        C::Error: Debug,
+    {
+        while let Some((req, expected_rsp)) = self.messages.next() {
+            // These unwraps could propagate errors with the correct
+            // bound on C::Error
+            let rsp = to_check.ready_and().await.unwrap().call(req).await.unwrap();
+            ensure!(
+                rsp == expected_rsp,
+                "Expected {:?}, got {:?}",
+                expected_rsp,
+                rsp
+            );
+        }
```

Comment:
:heart_eyes:

Reply:
Do you know what the correct bound on `C::Error` to be able to turn the `unwrap()` into a `?` is? Ideally this should work on any `Service`.

Reply:
Aah, so this is where we get fucked by tower requiring that we return `Box` from buffers. This is related to needing to put `.map_err(|e| eyre!(e))` on every callback from a tower service. The `eyre` macro uses autoref specialization to handle the fact that `Box<dyn Error>` isnt an error and its impossible to implement `From<Box<dyn Error>> for Report`. Ideally the bound would be `C::Error: Into<Report>`, but that doesn't work with `Box` which is the _only_ error type we actually use in practice .... 

_frustration_

So for now I'd say just require that the service is a `Box<dyn Error>` and put a `// TODO(jlusby): fix this once we fix tower error handling` when we'll be able to actually change the bound to `Into<Report>`